В банке дистанционной проверяющей системы имеется более 100000 заданий. Все задачи пронумерованы, начиная с единицы. Эти задания в течение учебного периода решают участники различных курсом. Каждому студенту при регистрации присваивается уникальный идентификатор – натуральное число, не превышающее 1000000. Студент может сдать несколько различных правильных решений одной задачи, при этом в зачёт идёт только одно из них.
Преподаватель сделал выгрузку результатов за некоторый период времени и выбрал студента, который решил наибольшее количество задач из банке с идущими подряд номерами, не пропустив ни одной.
Определите идентификационный номер студента, который решил наибольшее количество задач с идущими подряд номерами, и количество решённых им задач. Если несколько студентов решили одинаковое максимальное количество задач, то укажите наименьший идентификационный номер.
Входные данные
В первой строке входного файла находится число N – количество зачтённых решений (натуральное число, не превышающее 60000) за некоторый период времени. Каждая из следующих N строк содержит два натуральных числа, не превышающих 100000: идентификатор студента и номер правильно решённой задачи.
Выходные данные
Два целых неотрицательных числа: наименьший идентификационный номер студента и наибольшее количество решённых задач с подряд идущими номерами.
Типовой пример организации входных данных
9
40 3
60 33
60 33
50 125
50 126
50 127
40 4
50 72
50 126
Для приведённого примера студент с идентификационным номером 50 решил наибольшее количество задач с идущими подряд номерами (3 задачи). Ответом является пара чисел: 50; 3

In [1]:
file = open(r"D:\informaticsclass\egeinfo\26\26_19256.txt")
lines = sorted([list(map(int, i.strip().split())) for i in file])

best_students = []
best_streak = 0
curr = 1

for i in range(1, len(lines)):
    previous = lines[i - 1]
    current = lines[i]

    if previous[0] == current[0] and previous[1] == current[1]:
        continue

    if previous[0] == current[0] and previous[1] + 1 == current[1]:
        curr += 1
    else:
        if curr > best_streak:
            best_streak = curr
            best_students = [(previous[0], curr)]
        elif curr == best_streak:
            best_students.append((previous[0], curr))
        curr = 1

if curr > best_streak:
    best_streak = curr
    best_students = [(lines[-1][0], curr)]
elif curr == best_streak:
    best_students.append((lines[-1][0], curr))

print(*min(best_students, key=lambda x: x[0]))

40031 148


№ 23765 Демоверсия 2026 (Уровень: Базовый)

Отдел маркетинга сети магазинов составляет рейтинг продуктов по информации об их сроках хранения с момента изготовления и после
вскрытия упаковки. Для каждого продукта известен срок его хранения с момента изготовления и срок годности к употреблению после вскрытия упаковки. Продукты пронумерованы начиная с единицы.
В рейтинговом списке маркетологи располагают продукты по следующему алгоритму:
– все 2N чисел, обозначающих срок хранения и срок годности к употреблению для N продуктов, упорядочивают по возрастанию;
– если минимальное число в этом упорядоченном списке – срок хранения, то продукт в рейтинге занимает первое свободное место от
его начала;
– если минимальное число – срок годности к употреблению, то продукт занимает первое свободное место от конца рейтинга;
– если число обозначает срок хранения или срок годности к употреблению уже рассмотренного продукта, то его не принимают во внимание.
Этот алгоритм применяется последовательно для размещения всех N продуктов.
Определите номер последнего продукта, для которого будет определено его место в рейтинге, и количество продуктов, которые займут в рейтинге более низкие места.
Входные данные
В первой строке входного файла находится натуральное число N (N ≤ 1000) – количество продуктов. Следующие N строк содержат пары
чисел, обозначающих соответственно срок хранения продукта с момента изготовления и срок годности к употреблению после вскрытия упаковки (все числа натуральные, различные).
Запишите в ответе два натуральных числа: сначала номер последнего продукта, для которого будет определено его место в рейтинге, затем – количество продуктов, которые займут в рейтинге более низкие места. 

In [ ]:
file = open(r"D:\informaticsclass\egeinfo\26\26_23765.txt")
products = sorted([list(map(int, i.strip().split(" "))) for i in file])
last = products[-1]
first = []
second = []
rating = []



ValueError: [80263, 16927] is not in list

№ 23283 Основная волна 11.06.25 (Уровень: Базовый)

Входной файл содержит информацию о заявках граждан, обращающихся во многофункциональный центр (МФЦ) в течение календарных суток. В заявке указаны время начала и время окончания приёма специалистом (в минутах от начала суток).
Рабочие места специалистов МФЦ (окна) пронумерованы натуральными числами начиная с 1. Приём одного гражданина ведёт свободный специалист в окне с минимальным номером. Новый посетитель может обратиться к освободившемуся специалисту начиная со следующей минуты после завершения приёма предыдущего. Если в момент обращения в МФЦ свободных специалистов нет, то гражданин уходит. Определите, сколько граждан смогут попасть на приём в МФЦ в течение 24 ч, и каков номер окна специалиста, который начнёт принимать посетителя последним. Если таких окон несколько, укажите наименьший номер окна.
Входные данные
В первой строке входного файла находится натуральное число К, не превышающее 1000, - количество окон в МФЦ. Во второй строке
натуральное число N (N ≤ 10 000), обозначающее количество граждан. Каждая из следующих N строк содержит два натуральных числа, каждое из которых не превышает 1440: указанные в заявке время начала и время окончания приёма (в минутах от начала суток).
Запишите в ответе два числа: количество граждан, которые смогут воспользоваться услугами МФЦ, и номер окна, в котором специалист примет последнего гражданина.

Типовой пример организации данных во входном файле
2
5
30 60
40 100
59 60
61 100
101 144
При таких исходных данных воспользоваться услугами МФЦ смогут первый, второй, четвёртый и пятый граждане. Наименьший номер окна, где последний из граждан будет принят специалистом, - 1, так как будут свободны окна 1 и 2.

In [50]:
file = open(r"D:\informaticsclass\egeinfo\26\26_23283.txt")
sessions = [list(map(int, i.strip().split(" "))) for i in file]
session = sessions.sort()
# окон 267
windows = [0] * 267
acceptions = []

for session in sessions:
    start, end = session[0], session[1]
    for i in range(len(windows)):
        window = windows[i]
        if start > window:
            windows[i] = end
            acceptions.append(i + 1)
            break

print(len(acceptions), acceptions[-1])

793 2


№ 22605 (Уровень: Средний)

(Д. Бахтиев) Во время матча в расширенной версии игры «Морской бой» велась автоматическая фиксация попаданий по игровому полю. Каждое попадание фиксируется по координатам клетки игрового поля (целые положительные числа x и y) и времени попадания — целому числу t (в секундах от начала матча).

Игровое поле — прямоугольное. Координаты попаданий (x, y) указываются в формате: x — номер строки (считая сверху вниз, от 1 с шагом 1), y — номер столбца (слева направо, от 1 с шагом 1). Некоторые клетки могли быть поражены несколько раз.

Требуется определить такую клетку, в которую было произведено не менее двух попаданий, и среди них — найти ту, в которой минимальный интервал времени между любыми двумя её попаданиями — наименьший. В ответе нужно вывести сумму координаты этой клетки (x + y), а также этот минимальный интервал в секундах. Если таких клеток несколько, выберите среди них клетку с наименьшим номером строки, а если таких тоже несколько — с наименьшим номером столбца.

Входные данные:
В первой строке записано одно целое число N — количество попаданий (
1
≤
N
⩽
1000000
1≤N⩽1000000). В следующих N строках заданы тройки целых чисел x y t — координаты клетки и время попадания (
1
≤
x
⩽
10000
1≤x⩽10000, 
1
≤
y
⩽
10000
1≤y⩽10000, 
0
≤
t
≤
1
0
6
0≤t≤10 
6
 ).

Выходные данные:
Два целых числа: сумму номеров строки столбца подходящей клетки, затем минимальный интервал времени между двумя попаданиями в эту клетку.

Типовой пример организации входных данных:
7
2 3 10
1 1 3
2 3 15
3 2 20
1 1 10
2 3 17
1 1 6

Для приведённого примера более одного попаданий было в клетки с координатами (2, 3) и (1,1). Минимальный интервал между попаданиями, равный двум секундам, был у клетки с координатами (2, 3). Ответ: 5 2.

In [69]:
from collections import defaultdict

file = open(r"D:\informaticsclass\egeinfo\26\26_22605.txt")

matrix = defaultdict(list)

for line in file:
    line = list(map(int, line.strip().split()))
    x, y, time = line[0], line[1], line[2]
    matrix[(x, y)].append(time)


def find_min_gap(lst: list):
    if len(lst) < 2:
        return None
    lst.sort()
    mn = 10 ** 7
    for i in range(len(lst) - 1):
        mn = min(mn, lst[i + 1] - lst[i])
    return mn


res = []
for x, y in matrix:
    obj = matrix[(x, y)]
    gap = find_min_gap(obj)
    if gap is not None:
        res.append((gap, x, y))

ans = min(res)
print(ans[1] + ans[2], ans[0])

4552 53


№ 20910 Апробация 05.03.25 (Уровень: Средний)

При онлайн-покупке билета на концерт известно, какие места в зале уже заняты. Необходимо купить два билета на такие соседние места в одном ряду, чтобы перед ними все кресла с такими же номерами были свободны, а ряд находился как можно дальше от сцены. Если в этом ряду таких пар мест несколько, найдите пару с наименьшими номерами. В ответе запишите два целых числа: искомый номер ряда и наименьший номер места в найденной паре. Нумерация рядов и мест ведётся с 1. Гарантируется, что хотя бы одна такая пара в зале есть.

Входные данные

В первой строке входного файла находятся три числа: N – количество занятых мест в зале (целое положительное число,
не превышающее 10 000), M – количество рядов (целое положительное число, не превышающее 100 000) и K – количество мест в каждом ряду (целое положительное число, не превышающее 100 000). В следующих N строках находятся пары натуральных чисел: номер ряда и номер места занятого кресла соответственно (первое число не превышает значения M, а второе – K).

Выходные данные

Два целых положительных числа: наибольший номер ряда и наименьший номер места в найденной паре кресел.

Типовой пример организации данных во входном файле

7 7 8
1 1
6 6
5 5
6 7
4 4
2 2
3 3

При таких исходных данных ответом является пара чисел 5 и 6. Условию задачи удовлетворяют места 6 и 7 в ряду 5: перед креслами 6 и 7 нет занятых мест и это первая из двух возможных пар в этом ряду. В рядах 6 и 7 искомую пару найти нельзя.

In [8]:
file = open(r"D:\informaticsclass\egeinfo\26\26_20910.txt")
# 9992 99906 6661
matrix = [[] for _ in range(6661)]
for line in file:
    line = list(map(int, line.strip().split(" ")))
    horizontal, vertical = line[0] - 1, line[1] - 1
    matrix[vertical].append(horizontal)

res = []
for i_v in range(1, len(matrix)):
    try:
        left_seat = min(matrix[i_v - 1])
    except:
        left_seat = 99906 - 1
    try:
        right_seat = min(matrix[i_v])
    except:
        right_seat = 99906 - 1
    mn = min(left_seat, right_seat)
    if mn != 0:
        res.append([mn, i_v])

print(max(res, key=lambda x: x[0]))

[21028, 6660]


№ 9847 Основная волна 27.06.23 (Уровень: Сложный)

Система наблюдения ежеминутно фиксирует вход и выход посетителей магазина (в минутах, прошедших от начала суток). Считается, что в моменты фиксации входа и выхода посетитель находится в магазине. Нулевая минута соответствует моменту открытия магазина, который работает 24 ч в сутки без перерыва. Менеджер магазина анализирует данные системы наблюдения за прошедшие сутки, и выявляет отрезки времени наибольшей длины, в течение которых число посетителей, находящихся в магазине, не изменялось. Далее менеджер выбирает пики посещаемости — промежутки времени, когда количество посетителей в магазине было наибольшим. Пиков посещаемости в течение суток может быть несколько.

Входной файл содержит время входа и выхода каждого посетителя магазина. Определите, сколько пиков посещаемости было в течение суток, и укажите число посетителей в момент пика посещаемости.

Входные данные

В первой строке входного файла находится натуральное число N (N < 10000) - количество посетителей магазина.
Следующие N строк содержат пары чисел, обозначающих соответственно время входа и время выхода посетителя (все числа натуральные, не превышающие 1440).

Запишите в ответе два натуральных числа: сначала найденное количество пиков посещаемости, а затем число посетителей в момент пика посещаемости.

Типовой пример организации данных во входном файле

6
10 50
100 150
110 155
120 160
130 170
151 170

При таких исходных данных было два пика посещаемости: в отрезки времени со 130 по 150 минуты и со 151 по 155 минуты. Число посетителей в момент пика посещаемости равно 4. 

In [11]:
file = open(r"D:\informaticsclass\egeinfo\26\26_9847.txt")
# 987
lines = sorted([list(map(int, line.strip().split(" "))) for line in file])
history = []
curr = 0

for tick in range(1440):
    for guy in lines:
        enter, exit = guy[0], guy[1]
        if tick == enter:
            curr += 1
        if tick == exit:
            curr -= 1
    history.append(curr)

flag = False
most_time = []
c = 0
for i in range(len(history)):
    if history[i] == max(history):
        c += 1
        flag = True
    if history[i] != max(history) and flag:
        most_time.append(c)
        c = 0
        flag = False

print(len(most_time), max(history))

2 643


In [13]:
journal = [0 for i in range(24*60)]
with open(r"D:\informaticsclass\egeinfo\26\26_9847.txt") as f:
  n = f.readline()
  minutes = [list(map(int, x.split())) for x in f.readlines()]

for pair in minutes:
  for i in range(pair[0], pair[1] + 1):
    journal[i] += 1

m = max(journal)
for i in range(len(journal)):
  if journal[i] == m:
    print(i, journal[i])


739 643
740 643
741 643
742 643


№ 18185 (Уровень: Средний)

(Д. Бахтиев) На общегородском учебном совете было принято решение организовать поход классов на премьеру нового полнометражного фильма-триллера «ЭГЕ». Для недопущения массовых беспорядков было решено, что на один сеанс могут попасть ученики только одной параллели одной школы. Также нельзя разбивать одну параллель на несколько сеансов. Порядок, в котором школы и классы выбирают сеанс, строго регламентирован важной коллегией: приоритет отдаётся ученикам более старших классов (то есть сначала выбирают сеанс для учеников 11 класса, затем 10, затем 9 и т.д.) в порядке возрастания номеров школ. По известному количеству сеансов фильма и свободных мест для каждого сеанса определите, какое наибольшее количество параллелей смогут «насладиться» этим шедевром, а также количество учеников, которые увидят этот фильм.
В ответе запишите два целых числа — сначала количество классов, затем количество учеников в этих классах.

Входные данные
В первой строке входного файла находятся два числа, разделённые пробелом: число N, обозначающее общее количество разных параллелей (целое положительное число, не превышающее 10 000) и число K, обозначающее количество сеансов (целое положительное число, не превышающее 10 000). Каждая из следующих N строк содержит 3 числа через пробел: номер школы (целое положительное число, не превышающее 1000), номер класса для параллели (целое положительное число, не превышающее 11) и количество учеников в ней (целое положительное число, не превышающее 40). В последующих K строках находится одно число — количество свободных мест на очередном сеансе (целое неотрицательное число).

Выходные данные
Два натуральных числа: искомое количество параллелей и количество учеников в них.

Типовой пример организации данных во входном файле

3 2
15 10 22
899 11 35
44 3 25
27
26
При таких исходных данных на сеанс смогут пойти два класса: 10 класс (22 человека) из школы №15 и 3 класс (25 человек) из школы №44. Ответ: 2 47

In [27]:
file = open(r"D:\informaticsclass\egeinfo\26\26_18185.txt").readlines()
# 5555 3333
classes = [list(map(int, file[i].split(" "))) for i in range(5555)]
sessions = sorted([int(file[i]) for i in range(5555, 3333 + 5555)])

classes = sorted(classes, key=lambda x: [-x[1], x[0], x[2]])

students = 0
parallels = 0
for session in sessions:
    for i in range(len(classes)):
        if classes[i][2] <= session:
            students += classes[i][2]
            parallels += 1
            del classes[i]
            break

print(parallels, students)

3209 73122


№ 18541 (Уровень: Средний)

(В. Лашин) На престижном турнире по пауэрлифтингу 
M
M тяжелоатлетов соревнуются в силе, поднимая гирю.  Спортсмены поочередно подходят к стойке и выбирают вес для дальнейшего подъема. Каждый из них знает свои возможности и, стремясь к победе, выбирает один максимально возможный вес, который способен поднять, из предложенного на мероприятии набора из 
N
N разновесных снарядов.

После проведения турнира для получения статистических данных организаторы вычислили среднее значение весов всех снарядов, которые были выбраны атлетами, а также вес самого популярного снаряда. Определите, чему равны эти две величины.

Примечание. Гарантируется, что каждый атлет сможет выбрать для себя подходящий вес.

Входные данные:
Первая строка содержит два целых числа: 
N
N — количество доступных снарядов и 
M
M — количество атлетов  
(
1
≤
N
≤
50000
,
1
≤
M
≤
50000
)
(1≤N≤50000,1≤M≤50000). Следующие 
N
N строк содержат по одному целому числу — веса снарядов (от 1 до 100 000). Последние 
M
M строк содержат по одному целому числу — максимальные веса, которые могут поднять атлеты (от 1 до 100 000).
Выходные данные:
Запишите в ответе два целых числа — сначала целую часть среднего значения весов, которые выбрали атлеты, а затем вес снаряда, который был выбран максимальное количество раз. 
Типовой пример организации данных во входном файле 
3 3
50
100
70
60
80
65
При таких исходных данных первый и третий атлеты выберут вес 50, так как они не могут поднять 70 и 100. Второй атлет выберет вес 70, так как он не может поднять 100. Средний вес: (50 + 50 + 70) / 3 = 56. Чаще всего выбирали снаряд с весом 50. Ответ: 56 50.

In [ ]:
from collections import defaultdict
file = open(r"D:\Study\egeinfo\26\26_18541.txt").readlines()
# 45192 44910
weights = sorted([int(file[i]) for i in range(45192)])[::-1]
file = file[::-1]
athletes = [int(file[i]) for i in range(44910)]
usage = defaultdict(int)

for athlete in athletes:
    for weight in weights:
        if athlete >= weight:
            usage[weight] += 1
            break

ans = 0
for key, value in usage.items():
    ans += key * value

print(int(ans / 44910), max(usage.items(), key=lambda x: x[1])[0])

49989 65113


№ 9171 Джобс 06.06.2023 (Уровень: Сложный)

Поезд следует по магистрали через M населенных пунктов. Известно, что в поезде K мест. Дан список из N заявок на поездку, для каждой из которых известно, на какой станции пассажир собирается садиться, а на какой — выходить. При посадке на станции Х контроллер отдает предпочтение тому пассажиру, который едет дальше остальных, определяя место пассажира, как свободное с минимальным номером (от 1 до K). При этом сначала осуществляется высадка пассажиров, а затем посадка. 
Определите, сколько пассажиров смогут добраться до пункта своего назначения и сколько перегонов будут заняты все места поезда (перегон – участок магистрали между соседними населенными пунктами).
Входные данные:
В первой строке файла задано три числа: M (2 ≤ M ≤ 2000) – количество населенных пунктов со станциями на магистрали, K (1 ≤ K ≤ 1000) – количество мест в поезде и N (1 ≤ N ≤ 10000) – количество пассажиров, желающих проехать на поезде.
В каждой из последующих N строк располагаются пары чисел: сначала номер населенного пункта, откуда хочет начать свою поездку пассажир, затем номер населенного пункта, где пассажир собирается сойти с поезда.
Выходные данные:
Два числа: сначала количество пассажиров, которые смогут добраться до нужной им станции, затем количество перегонов, при прохождении которых в поезде будут заняты все места.
Пример входных данных:
10 3 6
2 6
2 4
3 5
3 8
4 9
4 6
При таких исходных данных добраться до нужного пункта смогут 4 пассажира ( (2, 6), (2, 4), (3, 8), (4, 9) ). При этом свободных мест не будет на перегонах 3 перегонах (3-4, 4-5 и 5-6). 

In [23]:
file = open(r"D:\informaticsclass\egeinfo\26\26_9171.txt")
passengers = sorted([list(map(int, line.split(" "))) for line in file], key=lambda x: x[0] - x[1])
# 1234 987 3333

k = 987
cnt = 0
full_seat = 0
arrived = 0

for station in range(1234 + 1):
    awaiting = []
    to_kill = []
    for i in range(len(passengers)):
        passenger = passengers[i]
        enter, exit = passenger[0], passenger[1]
        if station == enter: 
            awaiting.append((i, passenger))
        if station == exit:
            cnt -= 1
            arrived += 1
            to_kill.append(i)

    for person in awaiting:
        if cnt < k:
            cnt += 1
        else:
            to_kill.append(person[0])

    awaiting.clear()

    for i in sorted(set(to_kill), reverse=True):
        del passengers[i]

    if station < 1234 and cnt == k:
        full_seat += 1

print(arrived, full_seat)

2873 267


№ 8616 (Уровень: Сложный)

(Л. Шастин) В стоматологической клинике работает K специалистов. Все специалисты в базе данных сохранены по номерам от 1 до К.

В call-центр стоматологии звонят клиенты, желая записаться на прием ко врачу. В отчете предоставлена информация о звонках, которые происходили последовательно за определенный период времени. Известно время, на которое каждый клиент хочет записаться к специалисту, а также ID-номер специалиста, к которому хочет попасть клиент. При этом, согласно регламенту клиники, длительность любого приема составляет 30 минут. Администратор записывает клиента к специалисту, если на то время, в которое клиент желает пребывать на приёме, не назначено других, ранее сделанных записей. Но если запись к тому специалисту, к которому желает попасть клиент, невозможна, тогда администратор записывает клиента к специалисту с наименьшим ID-номером, среди всех тех, что свободны в рассматриваемое время. Специалисты могут принимать клиентов со следующей минуты после окончания приема предыдущего клиента. Если подходящих специалистов нет, то администратор просит прощения у клиента и сообщает, что он не может записать его.


Длительность рабочего дня стоматологической клиники составляет 840 минут. Последняя минута возможного начала приёма = 810.

Определите, сколько клиентов смогли записаться на прием, а также номер специалиста, к которому записался последний клиент.


Входные данные

В первой строке входного файла находится число N – количество клиентов, которые хотят записаться на прием (натуральное число, не превышающее 10000). Во второй строке находится число K – количество специалистов в стоматологической клинике. В следующих N строках находятся два значения: минута с которой клиент хочет записаться и номер специалиста, к которому клиент хочет записаться на прием. Отсчёт времени ведётся от начала рабочего дня стоматологии (все числа положительные, не превышающие 960). Данные в файле даны в том порядке, в котором клиенты звонили в call-центр стоматологии.

Запишите в ответе два целых числа: сначала количество клиентов, которые смогли записаться на прием, а затем номер специалиста, к которому записался последний клиент.


Типовой пример организации данных во входном файле

5
2
30 2
50 2
570 1
40 2
150 1

При таких исходных данных первый, второй, третий и пятый клиенты смогут записаться на прием. Последний клиент, которого смогли записать, был записан к первому специалисту.

Ответ для примера: 4 1.

Типовой пример имеет иллюстративный характер. Для выполнения задания используйте данные из прилагаемых файлов.

In [36]:
file = open(r"D:\informaticsclass\egeinfo\26\26_8616.txt")
# 10000
# 100
dentists = [-1 for i in range(100)]
clients = [list(map(int, line.split(" "))) for line in file]
success = []

for client in clients:
    time, id = client[0], client[1] - 1
    if time > 810: continue
    if dentists[id] < time:
        dentists[id] = time + 30
        success.append(id + 1)
    else:
        for i in range(len(dentists)):
            if dentists[i] < time:
                dentists[i] = time + 30
                success.append(i + 1)
                break
print(len(success), success[-1])

582 87


In [15]:
file = open(r"D:\informaticsclass\egeinfo\26\26_8616.txt")
dentists = {id: [0] * 841 for id in range(1, 101)}
clients = [list(map(int, line.split(" "))) for line in file]
success = []

for client in clients:
    time, id = client[0], client[1]
    if time > 810: continue
    if 1 not in dentists[id][time:time+30]:
        dentists[id][time:time+30] = [1] * 30
        success.append(id)
    else:
        for i in range(1, len(dentists) + 1):
            if 1 not in dentists[i][time:time+30]:
                dentists[i][time:time+30] = [1] * 30
                success.append(i)
                break
print(len(success), success[-1])

2105 96


№ 18030 (Уровень: Средний)

(Л. Шастин) Для построения магического карточного домика в стране грёз используется набор из N
N игральных карт разных мастей, имеющих весовые номера. Сам карточный домик состоит из некоторого количества уровней. Первый уровень состоит из наибольшего количества карт и служит опорой для остальных уровней. Каждый следующий уровень может включать в себя любое количество карт, которое меньше количества карт, из которых состоит предыдущий уровень. При этом сумма номеров карт, из которых состоит любой следующий уровень, должна быть строго меньше суммы номеров карт, из которых состоит текущий уровень. Идеальным считается карточный домик, состоящий из максимального количества уровней. Определите количество уровней в идеальном карточном домике, который можно построить из карт, имеющихся в данном наборе, а также минимально возможную сумму номеров всех карт, из которых может состоять такой карточный домик.

Входные данные 

В первой строке входного файла находится число 
N
N — количество карт в наборе (натуральное число, не превышающее 
10000
10000). В следующих 
N
N строках находятся номера карт (все числа натуральные, не превышающие 
100000
100000), каждое — в отдельной строке.

Запишите в ответе два целых числа: сначала количество уровней в идеальном карточном домике, затем минимально возможную сумму номеров всех карт, из которых может состоять такой карточный домик.

Типовой пример организации данных во входном файле 

8
2
9
8
4
12
2
10
3

Пример входного файла приведён для набора из восьми игральных карт. При таких исходных данных идеальный карточный домик будет состоять из трёх уровней: {4, 8, 9}, {3, 2}, {2}. Сумма номеров карт, из которых состоит этот домик, равна 28. Ответ: 3 28.

In [19]:
file = open(r"D:\informaticsclass\egeinfo\26\26_18030.txt")
# 9997
cards = sorted([int(i) for i in file])
level = 1
sum_cards = 0

while cards:
    if len(cards) < level: 
        level -= 1
        break
    for _ in range(level):
        card = cards.pop(0)
        sum_cards += card
    level += 1
print(level, sum_cards)

140 491509192


№ 8581 (Уровень: Средний)

(Л. Шастин) На склад магазина привезли N упаковок свежей продукции. Вновь привезенную продукцию сортируют по K холодильным камерам, вместимость каждой из которых равна M кг. Холодильные камеры, в свою очередь, пронумерованы от 1 до K. Фасовщики заполняют холодильные камеры последовательно, начиная с 1-й. Сначала погружают товары наибольшего объема (до тех пор, пока самый большой из оставшихся товаров влезает в холодильную камеру), стремясь заполнить текущую холодильную камеру до предела, а оставшееся свободное место начиняют товарами наименьшего объема. Гарантируется, что K камер хранения достаточно для сортировки всей продукции по описанной выше стратегии.

Необходимо определить номер холодильной камеры, в которую погрузили последний товар, а также остаток свободного в ней места.

Входные данные

В первой строке входного файла находится число N – количество упаковок привезенной продукции (натуральное число, не превышающее 5000). Во второй строке находится число K – количество холодильных камер. А в третьей строке находится число M - вместимость каждой из холодильных камер в кг. В следующих N строках находятся натуральные числа - веса упаковок в кг.

Запишите в ответе два целых числа: сначала номер холодильной камеры, в которую погрузили последний товар, а затем количество оставшегося в ней свободного места (в кг).

Типовой пример организации данных во входном файле

5
5
10
9
7
6
4
1

При таких исходных данных первая холодильная камера будет заполнена до отвала, во второй останется 3 кг свободного места, а в третьей - 0 кг. В третью же камеру и погрузят последний товар. Ответ: 3 0.

In [ ]:
file = open(r"D:\informaticsclass\egeinfo\26\26_8581.txt")
# 5000 – количество упаковок привезенной продукции
# 500 – количество холодильных камер
# 6000 - вместимость каждой из холодильных камер в кг
storage = [6000] * 500
packages = sorted([int(line) for line in file], reverse=True)

last = 0
left = 0
for idx, weight in enumerate(storage):
    curr_weight = weight
    while packages and curr_weight >= packages[0]:
        curr_weight -= packages[0]
        packages.pop(0)
    while packages and curr_weight >= packages[-1]:
        curr_weight -= packages[-1]
        packages.pop(-1) 
    if not packages:
        last = idx + 1
        left = curr_weight
        break
print(last, left)

В канун нового года на заводе по производству подарков Деда Мороза кипела работа, все конвейеры производили подарки. И вот, они наконец-то остановились, на складе Деда Мороза оказалось N подарков. Помощники Деда Мороза задали каждому подарку его привлекательность, находящуюся в диапазоне от 26 до 2000, а также соответствующую им категорию, заданную определённым словом. Всего существует 2 различные категории: еда(meal) и не еда(inedible). После этого необходимо было поделить все подарки на пары. Подарки делятся на пары, суммарная привлекательность подарков, составляющих пару, должна быть равна 2026. При этом всём, если подарки относятся к разным категориям, то радость ребёнка будет выше.  Помощники Деда Мороза потратили очень много времени на действия, описанные  выше, поэтому просят вас о помощи. Они предоставили всю информацию в файле 26.txt. Помогите им определить, какое максимальное количество подарков, которых равна 2026 можно собрать из имеющихся N подарков, а также какая максимальная радость может быть получена, если пара, состоящая из двух подарков, относящихся к разным категориям имеет радость равную 2026, а если к одной категории — 2025.

In [18]:
file = open(r"D:\informaticsclass\egeinfo\26\26 (1).txt")
# 992876
from collections import defaultdict
dc = {}
presents = []
happiness = 0
cnt = 0
for line in file:
    presents.append((int(line.strip().split(" ")[0]), line.strip().split(" ")[1][0]))
    try:
        dc[line.strip().split(" ")[0] + line.strip().split(" ")[1][0]] += 1
    except:
        dc[line.strip().split(" ")[0] + line.strip().split(" ")[1][0]] = 1

for present in presents:
    typ = present[1]
    if typ == "m":
        goal = "i"
    else:
        goal = "m"

    if dc[str(2026 - present[0]) + goal] != 0 and dc[str(present[0]) + typ] != 0:
        happiness += 2026
        dc[str(2026 - present[0]) + goal] -= 1
        dc[str(present[0]) + typ] -= 1
        cnt += 2

    elif present[0] == 1013 and dc[str(present[0]) + typ] >= 2:
        happiness += 2025
        dc[str(present[0]) + typ] -= 2
        cnt += 2

    elif dc[str(2026 - present[0]) + typ] != 0 and dc[str(present[0]) + typ] != 0:
        happiness += 2025
        dc[str(2026 - present[0]) + typ] -= 1
        dc[str(present[0]) + typ] -= 1
        cnt += 2
    else:
        continue

print(cnt, happiness)

4 4052
[(26, 'm'), (2000, 'i'), (1999, 'i'), (1013, 'i'), (1013, 'i'), (27, 'm'), (27, 'i')]
{'26m': 0, '2000i': 0, '1999i': 0, '1013i': 2, '27m': 0, '27i': 1}


Организация купила для своих сотрудников все места в нескольких рядах на стадионе. Известно, какие места уже распределены между сотрудниками. Найдите ряд с наименьшим номером, в котором есть четыре свободных соседних места, таких что слева и справа от них в том же ряду места уже распределены (заняты). Гарантируется, что есть хотя бы один ряд, удовлетворяющий условию. По заданной информации определите минимальный номер ряда, удовлетворяющего условию, и наибольший номер места из найденных в этом ряду четверок мест.

Входные и выходные данные

В первой строке входного файла находится одно число: N   — количество занятых мест (натуральное число, не превышающее 10000  ). В следующих N   строках находятся пары чисел: ряд и место выкупленного билета (числа не превышают 100   000  ).

Пример входных данных:

6  

4 1  

3 5  

4 6  

4 7  

3 10  

5 8  

При таких исходных данных ответ должен содержать 2   числа: 3   – минимальный номер ряда, и 9   – наибольший номер места из выбранной четверки в 3   ряду (была выбрана четверка с номерами 6,7,8,9  ).

In [33]:
file = open(r"D:\informaticsclass\egeinfo\26\Задание_26__mixo__t7qt.txt")
#10000
file = [list(map(int, line.strip().split(" "))) for line in file]
hall = [[] for i in range(max(file, key=lambda x: x[0])[0])]
for i in file:
    hall[i[0] - 1].append(i[1])
for i in range(len(hall)):
    hall[i] = sorted(hall[i])

for row, places in enumerate(hall):
    flag = False
    for i in range(len(places) - 1, 0, - 1):
        if places[i] - places[i - 1] == 5:
            print(row + 1, places[i] - 1)
            flag = True
            break
    if flag:
        break

2 17


В лесополосе осуществляется посадка деревьев. Причем саженцы высаживают рядами на одинаковом расстоянии.
Через какое-то время осуществляется аэросъемка, в результате которой определяется, какие саженцы прижились. Необходимо определить ряд с максимальным номером, в котором есть подряд ровно 7   неприжившихся саженцев, при условии, что справа и слева от них саженцы прижились. В ответе запишите сначала наибольший номер ряда, затем наименьший номер из неприжившихся мест.

Входные данные:
В первой строке входного файла находится число N   – количество прижившихся саженцев (натуральное число, не превышающее 10000  ). Каждая из следующих N   строк содержит два натуральных числа, не превышающих 100000  : номер ряда и номер занятого места.

Выходные данные:
Два целых неотрицательных числа, записанные через пробел: максимальный номер ряда, где нашлись обозначенные в задаче места, и минимальный номер подходящего места.

Пример входного файла:
7  
40 30  
40 34  
50 125  
50 129  
50 64  
50 68 
50 70  
Ответ для примера (при поиске 3   подряд идущих неприжившихся саженцах): 50 65  

In [1]:
file = open(r"D:\informaticsclass\egeinfo\26\Задание_26__o4nj__rs6l.txt")
# 10000
data = [list(map(int, line.strip().split(" "))) for line in file]
max_row = max(data, key=lambda x: x[0])[0]
rows = [[] for _ in range(max_row + 1)]

for row, place in data:
    rows[row].append(place)

for row in range(max_row, 0, -1):
    places = rows[row]
    if not places:
        continue
    places.sort()
    f = False
    for i in range(1, len(places)):
        if places[i] - places[i-1] == 8:
            print(row, places[i-1] + 1)
            f = True
            break
    if f:
        break

1536 134


При онлайн-покупке билета на концерт известно, какие места в зале уже заняты. Необходимо купить два билета на такие соседние места в одном ряду, чтобы перед ними было занято не более двух кресел с такими же номерами, а ряд находился как можно дальше от сцены. Сцена расположена перед первым рядом. Если в этом ряду таких пар мест несколько, найдите пару с наибольшими номерами мест. Нумерация рядов и мест ведётся с 1. Гарантируется, что хотя бы одна такая пара в зале есть. Определите наибольший номер ряда и наибольший номер места для найденной пары мест.

Входные данные представлены в файле следующим образом. В первой строке входного файла находятся три числа: N   – количество занятых мест в зале (целое положительное число, не превышающее 10000), M   – количество рядов (целое положительное число, не превышающее 100 000) и K   – количество, мест в каждом ряду (целое положительное число, не превышающее 100 000). В следующих N   строках находятся пары натуральных чисел: номер ряда и номер места занятого кресла соответственно (первое число не превышает значения M  , а второе – K  ).

Запишите в ответе два целых числа: искомый номер ряда и наибольший номер места в найденной паре.

Пример входного файла:
8 7 4
1 1
4 1
3 2
1 3
2 3
6 3
5 4
7 2
Условию задачи удовлетворяют места 1 и 2 в ряду 6: перед 1 креслом 2 занятых места, а перед 2 местом 1 занятое место.

In [5]:
# Открываем файл с данными
f = open(r"D:\informaticsclass\egeinfo\26\26__86lva (1).txt")

# Читаем основные параметры:
# n - количество занятых мест
# m - количество рядов
# k - количество мест в каждом ряду
n, m, k = map(int, f.readline().split())

# Создаем структуру для хранения занятых мест по рядам
# row[i] будет содержать список занятых мест в ряду i+1
# Изначально добавляем m+1 как маркер конца ряда
row = [[m+1] for i in range(k)]

# Читаем информацию о занятых местах
for i in range(n):
    x, y = map(int, f.readline().split())
    row[y-1].append(x) # Добавляем занятое место в соответствующий ряд

# Инициализируем переменные для поиска оптимального решения
mx_r = 0 # Максимальный номер ряда (как можно дальше от сцены)
mx_i = 0 # Максимальный номер места

# Проходим по всем парам соседних рядов
for i in range(len(row) - 1):
    # Находим максимальное занятое место из двух соседних рядов
    # Чем больше это значение, тем дальше от сцены мы можем сесть
    if max(row[i + 1] + row[i]) >= mx_r:
        x = max(row[i + 1] + row[i])

        # Ищем подходящее место, удовлетворяющее условиям:
        # - место не должно быть занято в текущем ряду
        # - место не должно быть занято в соседнем ряду
        # - перед этим местом должно быть не более 2 занятых мест
        while (x in row[i] or x in row[i + 1]
               or len([p for p in row[i] if p < x]) > 2
               or len([p for p in row[i + 1] if p < x]) > 2):
            x -= 1 # Уменьшаем номер места, если условия не выполняются

        # Если нашли подходящее место дальше от сцены, чем предыдущие
        if x >= mx_r:
            mx_r = x # Сохраняем новый максимальный номер ряда
                                                                                                  
                                                                                                  
            mx_i = i # Сохраняем индекс ряда

# Выводим результат: номер ряда и номер места
print(mx_r, mx_i+2)

4484 356


В магазине есть N камер хранения. Номера камер начинаются с 1. В магазин приходят покупатели и оставляют свои вещи в камерах хранениях. Покупатель занимает одну из камер с минимальным номером. Следующий покупатель может занять камеру хранения со следующей минуты после освобождения камеры. Известно время сдачи вещей каждого покупателя и время выдачи вещей каждому покупателю. Если свободных камер нет, то покупатели уходят. Если один момент пришло несколько покупателей, то сначала вещи в камеру положит клиент, у которого большее количество времени вещи будут находиться в камере.

Входные данные: В первой строке файла находится натуральное число K - количество покупателей. Во второй строке файла находится натуральное число N– количество камер хранения в магазине. В последующих строках записано по два числа, не превышающих 1500: время прибытия клиента, время ухода клиента.

Запишите в ответ два числа через пробел: количество покупателей, которые смогли положить вещи в камеру хранения, и номер камеры, в которую положил вещи сотый покупатель.

In [7]:
file = open(r"D:\informaticsclass\egeinfo\26\26_9__3ck5k (1).txt")
# 1111
# 57
file = sorted([list(map(int, line.split(" "))) for line in file], key=lambda x: [x[0], -abs(x[0] - x[1])])
boxes = [[] for i in range(57)]

cnt = 0
i_box = 0
for i, person in enumerate(file):
    start, stop = person
    for idx, box in enumerate(boxes):
        if not box or start > box:
            boxes[idx] = stop
            cnt += 1
            if cnt == 100:
                i_box = idx + 1
            break
print(cnt, i_box)

256 4


Входной файл содержит сведения о заявках на проведение мероприятий в конференц-зале. В каждой заявке указаны время начала и время окончания мероприятия (в минутах от начала суток). Если время начала одного мероприятия меньше времени окончания другого, то провести можно только одно из них. Если время окончания одного мероприятия совпадает со временем начала другого, то провести можно оба. Определите, какое максимальное количество мероприятий можно провести в конференц-зале и каков при этом максимальный возможный перерыв между двумя последними мероприятиями.

Входные данные.

В первой строке входного файла находится натуральное число N — количество заявок на проведение мероприятий (натуральное число, не превыщающее 1000).

Следующие N строк содержат пары чисел, обозначающих время начала и время окончания мероприятия (каждое из чисел натуральное, не превосходящее 1440).

Запишите в ответе два числа через пробел: максимальное количество мероприятий и самый длинный перерыв между двумя последними мероприятиями (в минутах).

In [21]:
file = open(r"D:\informaticsclass\egeinfo\26\1_26_conf__3uznj.txt")
# 966
cnt = 1
file = sorted([list(map(int, line.split(" "))) for line in file], key=lambda x: x[1])
conf = [file.pop(0)]

for idx, meet in enumerate(file):
    start, stop = meet
    if start >= conf[-1][1]:
        conf.append(meet)
        cnt += 1
last = max([i for i in file if i not in conf])
print(cnt, 1299 - 1273)

40 26


На производстве штучных изделий N деталей должны быть отшлифованы и окрашены. Для каждой детали известно время её шлифовки и время окрашивания. Детали пронумерованы начиная с единицы. Параллельная обработка деталей не предусмотрена. На ленте транспортёра имеется N мест для каждой из N деталей. Места для деталей пронумерованы начиная с единицы. На ленте транспортёра детали располагают по следующему алгоритму:

- все 2N чисел, обозначающих время окрашивания и шлифовки для N деталей, упорядочивают по возрастанию;

- если минимальное число в этом упорядоченном списке — это время шлифовки конкретной детали, то деталь размещают на ленте транспортёра на первое свободное место от её начала;

- если минимальное число — это время окрашивания, то деталь размещают на первое свободное место от конца ленты транспортёра;

Этот алгоритм применяется последовательно для размещения всех N деталей. Определите сколько деталей будет отшлифовано, и деталь с каким номером окажется на позиции с номером K на ленте транспортёра.

В первой строке входного файла находится натуральное число N (N < 1000) – количество деталей и натуральное число K (K ≤  N). Следующие N строк содержат пары чисел, обозначающих соответственно время шлифовки и время окрашивания конкретной детали (все числа натуральные, различные).

Запишите в ответе два натуральных числа через пробел: сначала сколько деталей будет отшлифовано, затем номер детали, которая окажется на позиции c номером K на ленте транспортёра.

In [ ]:
# тут какая то непонятная возня с индексами в конце
file = open(r"D:\informaticsclass\egeinfo\26\26_2M__3whpt.txt").readlines()
left_side = []
right_side = []
# 997 333
k = 333
for idx, line in enumerate(file):
    line = list(map(int, line.split(" ")))
    a, b = line
    if a == min(a, b):
        left_side.append((idx, line))
    else:
        right_side.append((idx, line))

left_side.sort(key=lambda x: x[1])
right_side.sort(key=lambda x: x[1], reverse=True)
total = left_side + right_side

print(len(left_side), total[k - 1][0])
print(len(left_side), total[k][0])
# это правильный ответ
print(len(left_side), total[k - 1][0] + 1)



# решение школково
file = open("26_12933.txt")
# Kоличество деталей и номер позиции
n, k = map(int, file.readline().split())
line_start = [] # Список деталей в начале ленты для шлифовки
line_end = [] # Список деталей в конце ленты для окраски
c = 0 # Номер детали
for i in file:
    c += 1 # Увеличиваем номер детали
    t1, t2 = map(int, i.split()) # Время шлифовки и время окрашивания
    if t1 < t2: # Если время шлифовки меньше времени окрашивания
        line_start.append([t1, c]) # Добавляем деталь в начало ленты
    else: # Если время окрашивания не больше времени шлифовки
        line_end.append([t2, c]) # Добавляем деталь в конец ленты
line_start.sort() # Сортируем список шлифовки по возрастанию
line_end.sort(reverse=True) # Сортируем список окрашивания по убыванию
res = line_start + line_end # Объединяем списки
# Количество отшлифованных деталей
print(len(line_start))
print(res[k-1][1]) # Номер детали на позиции c номером K на ленте

На вокзале есть N   камер хранения для пассажиров, которые пронумерованы с единицы. Приходя на вокзал, пассажир кладет свои вещи в свободную камеру с минимальным номером. 
Известно время, когда пассажиры сдают и забирают вещи (в минутах с начала суток). 
Камера доступна для багажа начиная со следующей минуты после окончания срока хранения. 
В случае, когда свободных камер не находится, багаж не принимается и пассажир уходит. 
Если два пассажира пришли одновременно, то первым положит вещи тот, кто быстрее освободит ячейку камеры.

Определите пару пассажиров, которым удалось оставить вещи в камерах и у которых суммарное время хранения багажа максимально и кратно 53.

Входные данные. В первой строке входного файла находится число N   – количество камер хранения (натуральное число, не превышающее 1000), во второй строке число K   – количество пассажиров, сдающих багаж (натуральное число, не превышающее 10000). Каждая из следующих N   строк содержит два натуральных числа, не превышающих 1440: время сдачи багажа и время выдачи багажа.

Выходные данные. Программа должна вывести одно число: максимальное суммарное время хранения багажа двух пассажиров, кратное 53.

In [10]:
file = open(r"D:\informaticsclass\egeinfo\26\2-26__6dos6.txt")
n = int(file.readline())
k = int(file.readline())
file = sorted([list(map(int, line.strip().split(" "))) for line in file])
cells = [[] for i in range(n)]
accepted = []

for person in file:
    start, stop = person
    for idx, cell in enumerate(cells):
        if not cell or start > cell[1]:
            cells[idx] = person
            accepted.append(abs(start - stop))
            break

dc = {}
for i in accepted:
    try:
        dc[i % 53] += [i]
    except:
        dc[i % 53] = [i]

mx = 0
cnt = 0
for c in range(53):
    try:
        a = max(dc[c]) + max(dc[53 - c])
        mx = max(mx, a)
    except Exception as e:
        cnt += 1

print(mx)

1908


Входной файл содержит заявки пассажиров, желающих сдать свой багаж в камеру хранения. В заявке указаны время сдачи багажа и время освобождения ячейки (в минутах от начала суток). Багаж одного пассажира размещается в одной свободной ячейке с минимальным номером. Ячейки пронумерованы начиная с единицы. Освобождение ячейки происходит в течение 1 мин, освобождение начинается в минуту ухода клиента. Багаж можно поместить в освобождённую ячейку сразу после освобождения. Если в момент сдачи багажа свободных ячеек нет, то пассажир уходит.

Входные данные представлены в файле следующим образом.
В первой строке входного файла находится натуральное число K  , не превышающее 1000, – количество ячеек в камере хранения. Во второй строке – натуральное число N   (N   ≤  1000), обозначающее количество пассажиров. Каждая из следующих N   строк содержит два натуральных числа, каждое из которых не превышает 1440: указанное в заявке время размещения багажа в ячейке и время освобождения ячейки (в минутах от начала суток).
Требуется выбрать для каждой ячейки один багаж, который удалось оставить в данной ячейке так, чтобы сумма времени хранения всех выбранных багажей была минимальная и четная.

В ответе запишите одно число: найденную сумму.
Пример входного файла:
2
5
30 60
40 1000
59 60
61 1000
1010 1440

In [5]:
file = open(r"D:\informaticsclass\egeinfo\26\1-26__6dopt.txt")
n = int(file.readline())
k = int(file.readline())
file = sorted([list(map(int, line.strip().split(" "))) for line in file])
cells = [[] for i in range(n)]
diff = [[] for i in range(n)]

for person in file:
    start, stop = person
    for idx, cell in enumerate(cells):
        if not cell or start > cell[-1][1]:
            cells[idx].append(person)
            diff[idx].append(abs(start - stop))
            diff[idx].sort()
            break

total = sum([min(i) for i in diff])
changes = [[lst[i] - lst[0] for i in range(1, len(lst))] for lst in diff]
mn = 10 ** 7
for i in changes:
    for y in i:
        if y % 2 == 1:
            mn = min(mn, y)
print(total + mn)

213128


На вокзале есть N   камер хранения для пассажиров, которые пронумерованы с единицы. Приходя на вокзал, пассажир кладет свои вещи в свободную камеру с минимальным номером. Известно время, когда пассажиры сдают и забирают вещи (в секундах с начала суток). Камера доступна для багажа начиная со следующей секунды после окончания срока хранения. В случае, когда свободных камер не находится, багаж не принимается и пассажир уходит.

Определите пару пассажиров, которым удалось оставить вещи в камерах хранения и у которых суммарное время хранения багажа максимально и кратно 47.

Входные данные. В первой строке входного файла находится число K   – количество пассажиров, сдающих багаж (натуральное число, не превышающее 10000), во второй строке число N   – количество камер хранения (натуральное число, не превышающее 1000). Каждая из следующих K   строк содержит два натуральных числа, не превышающих 86400: время сдачи багажа и время выдачи багажа.

Выходные данные. Программа должна вывести одно число: максимальное суммарное время хранения багажа двух пассажиров, кратное 47.

In [8]:
file = open(r"D:\informaticsclass\egeinfo\26\3-26__68ecw.txt")
n = int(file.readline())
k = int(file.readline())
file = sorted([list(map(int, line.strip().split(" "))) for line in file])
cells = [[] for i in range(n)]
accepted = []

for person in file:
    start, stop = person
    for idx, cell in enumerate(cells):
        if not cell or start > cell[1]:
            cells[idx] = person
            accepted.append(abs(start - stop))
            break

dc = {}
for i in accepted:
    try:
        dc[i % 47] += [i]
    except:
        dc[i % 47] = [i]

mx = 0
cnt = 0
for c in range(47):
    try:
        a = max(dc[c]) + max(dc[47 - c])
        mx = max(mx, a)
    except Exception as e:
        cnt += 1

print(mx)

166192


Входной файл содержит информацию о заказах на использование фотостудии в течение суток. Каждый заказ включает время начала и окончания съёмки (в минутах от начала суток), а также сумму, которую клиент готов заплатить за аренду студии. Съемки можно проводить только со следующей минуты после окончания предыдущей.
Определите максимальный доход, который можно получить за аренду студии, а также общую продолжительность съёмок, которые будут проведены в этом случае. Если максимальный доход можно получить несколькими способами, то в ответ нужно записать минимальную общую продолжительность съёмок, при которой будет получен максимальный доход.

Входные данные
В первой строке файла записано натуральное число N   (1 ≤ N ≤ 1000  ) — количество заказов. Каждая из следующих N строк содержит по три числа: время начала, время окончания съёмки (натуральные числа, не превышающие 1440  ), и сумму оплаты.

Выходные данные
Два числа через пробел: максимальный доход и суммарная продолжительность выбранных съёмок.

Пример входного файла:
5  
50   100   2  
90   130   4  
110   140   5  
135   180   3  
170   210   6  
Максимальный доход достигается при выборе заказов 1  , 3  , 5  . Общая продолжительность: (100− 50 +1) +(140− 110 + 1) + (210 − 170 + 1) = 123  . Ответ: 13   123  .

In [ ]:
file = open(r"D:\informaticsclass\egeinfo\26\26_1__6k161.txt")
n = int(file.readline())
file = sorted([list(map(int, line.split(" "))) for line in file], key=lambda x: x[1])
price = [i[-1] for i in file]
duration = [abs(i[0] - i[1]) + 1 for i in file]

for i in range(len(file)):
    for j in range(i):
        # если мы можем провести съемку j до текущей съемки
        if file[i][0] > file[j][1]:
            # новые цена и время
            new_price = file[i][2] + price[j]
            new_duration = (file[i][1] - file[i][0] + 1) + duration[j]

            # если стоимость больше по-любому берем новые значения тк приоритет по цене
            if new_price > price[i]:
                price[i] = new_price
                duration[i] = new_duration
            # если стоимость такая же, а время меньше берем время поменьше
            elif new_price == price[i] and new_duration < duration[i]:
                duration[i] = new_duration
                    
res = sorted([(price[i], duration[i]) for i in range(n)], key=lambda x: (-x[0], x[1]))
print(*res[0])

593 971


На производстве штучных изделий N деталей должны быть отшлифованы и окрашены. Для каждой детали известно время её шлифовки и время окрашивания. Детали пронумерованы начиная с единицы. Параллельная обработка деталей не предусмотрена. На ленте транспортёра имеется N мест для каждой из N деталей. На ленте транспортёра детали располагают по следующему алгоритму:
— все 2N чисел, обозначающих время окрашивания и шлифовки для N деталей, упорядочивают по возрастанию;
— если минимальное число в этом упорядоченном списке — это время шлифовки конкретной детали, то деталь размещают на ленте транспортёра на первое свободное место от её начала;
— если минимальное число — это время окрашивания, то деталь размещают на первое свободное место от конца ленты транспортёра.
— если число обозначает время окрашивания или шлифовки уже рассмотренной детали, то его не принимают во внимание.
Этот алгоритм применяется последовательно для размещения всех N деталей. Определите номер предпоследней детали, для которой будет определено её место на ленте транспортёра, и количество деталей, которые будут покрашены до неё.

Входные данные:
В первой строке входного файла находится натуральное число N (N < 1000) – количество деталей. Следующие N строк содержат пары чисел, обозначающих соответственно время шлифовки и время окрашивания конкретной детали (все числа натуральные, различные).
Запишите в ответе два натуральных числа через пробел: сначала номер предпоследней детали, для которой будет определено её место на ленте транспортёра, затем количество деталей, которые будут покрашены до нее.

In [ ]:
file = open(r"D:\informaticsclass\egeinfo\26\26_1M__3whph.txt")
n = int(file.readline())

left = []
right = []
order = []   # порядок, в котором определяется месцто детали

for i in range(1, n + 1):
    x, y = map(int, file.readline().split())

    if x < y:
        left.append((x, i))
        order.append((x, i))
    else:
        right.append((y, i))
        order.append((y, i))

# итоговый порядок на ленте
total = sorted(left) + sorted(right, reverse=True)
belt = [i for t, i in total]

# порядок рассмотрения деталей
order.sort()
pred_posled = order[-2][1]

# сколько деталей будет покрашено до неё
painted_before = belt.index(pred_posled)

print(pred_posled, painted_before)

798 487


На производстве штучных изделий N деталей должны быть отшлифованы и окрашены. Для каждой детали известно время ее шлифовки и время окрашивания. Детали пронумерованы начиная с единицы. Параллельная обработка деталей не предусмотрена. На ленте транспортера имеется N мест для каждой из N деталей. На ленте транспортера детали располагают по следующему алгоритму:
– все 2N чисел, обозначающих время окрашивания и шлифовки для N деталей, упорядочивают по возрастанию;
– если минимальное число в этом упорядоченном списке – это время шлифовки конкретной детали, то деталь размещают на ленте транспортера на первое свободное место от ее начала;
– если минимальное число – это время окрашивания, то деталь размещают на первое свободное место от конца ленты транспортера;
– если число обозначает время окрашивания или шлифовки уже рассмотренной детали, то его не принимают во внимание.
Этот алгоритм применяется последовательно для размещения всех N деталей.
Определите сколько деталей будет отшлифовано и какой номер будет иметь деталь, которая была размещена на ленте последней.

Входные данные представлены в файле 26-1.txt следующим образом. Первая строка входного файла содержит натуральное число N (1 ≤ N ≤ 1000)   – количество деталей. Следующие N строк содержат пары чисел, обозначающих соответственно время шлифовки и время окрашивания конкретной детали (все числа натуральные, различные).
Запишите в ответе два натуральных числа через пробел: сначала количество деталей, которые будут отшлифованы, затем номер последней размещенной детали.

In [25]:
file = open(r"D:\informaticsclass\egeinfo\26\26_1__3whs4 (1).txt")
n = int(file.readline())
left = []
right = []
for i in range(1, n + 1):
    x, y = map(int, file.readline().split())
    if x < y:
        left.append((x, i))
    else:
        right.append((y, i))

total = sorted(left) + sorted(right, reverse=True)
left = sorted(left)
right = sorted(right, reverse=True)

if not right or (left and left[-1][0] > right[0][0]):
    last = left[-1][1]
else:
    last = right[0][1]

print(len(left), last)

484 544


Система наблюдения ежеминутно фиксирует вход и выход посетителей магазина (в минутах, прошедших от начала суток). Считается, что в моменты фиксации входа и выхода посетитель находится в магазине. Нулевая минута соответствует моменту открытия магазина, который работает 24 ч в сутки без перерыва. Менеджер магазина анализирует данные системы наблюдения за прошедшие сутки, и выявляет отрезки времени наибольшей длины, в течение которых число посетителей, находящихся в магазине, не изменялось. Далее менеджер выбирает пики посещаемости – промежутки времени, когда количество посетителей в магазине было наибольшим. Пиков посещаемости в течение суток может быть несколько.

Входной файл содержит время входа и выхода каждого посетителя магазина. Определите, сколько пиков посещаемости было в течение суток, и укажите число посетителей в момент пика посещаемости. Входные данные представлены в файле 26_6.txt следующим образом. Первая строка входного файла содержит натуральное число N (1 ≤ N ≤ 10000)   – количество посетителей магазина. Следующие N строк содержат пары чисел, обозначающих соответственно время входа и время выхода посетителя (все числа натуральные, не превышающие 1440).

Запишите в ответе два натуральных числа: сначала найденное количество пиков посещаемости, а затем число посетителей в момент пика посещаемости.
Примечание: пиком в этой задаче назовём глобальный максимум.

In [1]:
file = open(r"D:\informaticsclass\egeinfo\26\26_6__4103i.txt")
n = int(file.readline())
minutes = [0 for i in range(1440)]
for human in file:
    enter, exit = list(map(int, human.split(" ")))
    for i in range(enter, exit + 1):
        minutes[i] += 1

a = ""
for i in minutes:
    if i != max(minutes):
        a += " "
    else:
        a += "*"

print(len(a.split()), max(minutes))

1 3716


В магазине для упаковки подарков есть N кубических коробок. Самой интересной считается упаковка подарка по принципу матрёшки – подарок упаковывается в одну из коробок, та в свою очередь в другую коробку и т.д. Одну коробку можно поместить в другую, если длина её стороны хотя бы на 3 единицы меньше длины стороны другой коробки. Определите наибольшее количество коробок, которое можно использовать для упаковки одного подарка, и максимально возможную длину стороны самой маленькой коробки, где будет находиться подарок. Размер подарка позволяет поместить его в самую маленькую коробку.

Входные данные
В первой строке входного файла находится число N   – количество коробок в магазине (натуральное число, не превышающее 10 000). В следующих N   строках находятся значения длин сторон коробок (все числа натуральные, не превышающие 10 000), каждое – в отдельной строке. Запишите в ответе два целых числа: сначала наибольшее количество коробок, которое можно использовать для упаковки одного подарка, затем максимально возможную длину стороны самой маленькой коробки в таком наборе.

Типовой пример организации данных во входном файле
5
43
40
32
40
30
Пример входного файла приведён для пяти коробок и случая, когда минимальная допустимая разница между длинами сторон коробок, подходящих для упаковки «матрёшкой», составляет 3 единицы. При таких исходных данных условию задачи удовлетворяют наборы коробок с длинами сторон 30, 40 и 43 или 32, 40 и 43 соответственно, т.е. количество коробок равно 3, а длина стороны самой маленькой коробки равна 32.

In [4]:
file = open(r"D:\informaticsclass\egeinfo\26\26__7a6rm.txt")
n = int(file.readline())
boxes = sorted([int(line) for line in file], reverse=True)
accepted = [boxes.pop(0)]
for i in boxes:
    if i + 3 <= accepted[-1]:
        accepted.append(i)

print(len(accepted), accepted[-1])

2767 51


Для обеспечения безопасности на шоссе устанавливаются камеры видеонаблюдения. Установка возможна только в заранее определённых точках — на мачтах вдоль дороги. Известны расстояния от начала шоссе до каждой мачты.
По нормативам, чтобы камеры не мешали друг другу и не дублировали зоны наблюдения, любые две установленные камеры должны находиться на расстоянии не менее 10   метров друг от друга.
Необходимо установить максимально возможное количество камер, не нарушая нормативов и определить максимально возможное расстояние от начала шоссе до ближайшей установленной камеры при этом размещении.

Входные данные
В первой строке — число N   (натуральное, не больше 10000  ) — количество мачт. В следующих N   строках — расстояния от начала шоссе до каждой мачты (натуральные числа не больше 10000  ), каждое в отдельной строке.
Выходные данные
Два целых числа: максимальное количество установленных камер и максимально возможное расстояние от начала шоссе до ближайшей установленной камеры
Пример входного файла:
8  
15  
25  
5  
35  
12  
45  
20  
3  
При таких входных данных мы можем установить камеры на мачты 45  , 35  , 25  , 15  , 5   или 2  , 12  , 25  , 35  , 45  . Максимальное количество камер - 5, а максимальное расстояние до ближайшей камеры - 5.

In [7]:
file = open(r"D:\informaticsclass\egeinfo\26\26_2__6gp6f (1).txt")
n = int(file.readline())
cameras = sorted([int(line) for line in file], reverse=True)
accepted = [cameras.pop(0)]

for camera in cameras:
    if accepted[-1] - camera >= 10:
        accepted.append(camera)

print(len(accepted), accepted[-1])

946 2


Для освещения туристической тропы необходимо установить фонари. Установка фонаря возможна только в заранее определённых точках вдоль тропы — на специальных столбах. Известно расстояние от начала тропы до каждого такого столба.
По техническим требованиям, чтобы избежать перекрытия световых зон и сэкономить электроэнергию, любые два соседних фонаря должны находиться на расстоянии не менее 8   метров друг от друга.
Администрация заказала установить максимально возможное количество фонарей, не нарушая требований, а также хочет узнать, максимально возможное расстояние от начала тропы до ближайшего установленного фонаря при таком размещении.

Входные данные
В первой строке входного файла находится число N   (натуральное, не превышает 10000  ) — количество столбов, на которые можно установить фонари. В следующих N   строках — расстояния (в метрах) от начала тропы до каждого столба (все числа натуральные, не превышающие 10000  ), каждое в отдельной строке.

Выходные данные
В ответе два целых числа: максимально возможное количество фонарей и максимально возможное расстояние от начала тропы до ближайшего установленного фонаря при таком размещении.

Пример входного файла:
6  
10  
17  
26  
19  
5  
30  
При таких исходных данных мы можем установить фанарь на столбы 5  , 17  , 26   или на 30  , 19  , 10  . Максимальное количество фонарей - 3  , а максимальное расстояние до ближайшего фонаря - 10  . Ответ: 3   10  .

In [9]:
file = open(r"D:\informaticsclass\egeinfo\26\26_1__6gp5h.txt")
n = int(file.readline())
cameras = sorted([int(line) for line in file], reverse=True)
accepted = [cameras.pop(0)]

for camera in cameras:
    if accepted[-1] - camera >= 8:
        accepted.append(camera)

print(len(accepted), accepted[-1])

1214 6


Входной файл содержит сведения о заявках на бронирование помещения для проведения консультаций. Каждая заявка содержит время начала и окончания консультации (в минутах от начала суток), а также сумму, которую готовы заплатить за проведение консультации. Два мероприятия можно провести, если время окончания одного меньше или равно времени начала другого.
Определите максимальную сумму оплаты, которую можно получить за день, а также общую длительность всех выбранных консультаций. Если максимальный доход можно получить несколькими способами, то в ответ нужно записать минимальную общую продолжительность консультаций, при которой будет получен максимальный доход.

Входные данные:
В первой строке входного файла — натуральное число N   (1 ≤ N ≤ 1000  ) —количество заявок. В каждой из следующих N   строк записаны три числа: время начала и окончания консультации (в минутах от начала суток, от    1   до 1440  ) и сумма оплаты.
Выходные данные:
Два числа через пробел: максимальную сумму оплаты за день и суммарную длительность всех выбранных консультаций.
Пример входного файла:

5  
10   60   5  
60   100   6  
110   150   7  
160   200   8  
100   180   15  
Максимальный доход достигается при выборе заявок 1  , 2  , 5  . Общая продолжительность: (60− 10 +1) +(100− 60 + 1)+ (180 − 100 + 1) = 173  . Ответ: 26   173  .

In [ ]:
file = open(r"D:\informaticsclass\egeinfo\26\26_3__6wzob.txt")
n = int(file.readline())
conf = sorted([list(map(int, line.split(" "))) for line in file], key=lambda x: [x[1], x[2], x[0]])
price = [i[2] for i in conf]
# очень важно прибавлять к разнице единицу!!!!!
time = [i[1] - i[0] + 1 for i in conf]

for i in range(len(conf)):
    for j in range(i):
        if conf[i][0] >= conf[j][1]:
            new_price = conf[i][2] + price[j]
            # очень важно прибавлять к разнице единицу!!!!!
            new_time = (conf[i][1] - conf[i][0] + 1) + time[j]

            if new_price > price[i]:
                price[i] = new_price
                time[i] = new_time
            elif new_price == price[i] and new_time < time[i]:
                time[i] = new_time

res = sorted([(price[i], time[i]) for i in range(n)], key=lambda x: [-x[0], x[1]])[0]
print(*res)

629 968
2
[[5, 12, 6], [6, 12, 8], [14, 15, 8], [12, 15, 9], [17, 36, 5], [24, 57, 9], [3, 60, 9], [16, 62, 1], [2, 63, 4], [55, 66, 8], [47, 72, 4], [45, 75, 6], [27, 77, 1], [35, 78, 8], [18, 80, 8], [71, 82, 7], [74, 82, 9], [40, 83, 4], [83, 85, 2], [46, 86, 4], [64, 86, 4], [59, 86, 6], [1, 86, 8], [29, 87, 5], [60, 93, 2], [84, 95, 4], [46, 96, 8], [46, 98, 10], [26, 99, 8], [34, 99, 8], [4, 100, 5], [26, 100, 8], [20, 101, 1], [18, 101, 3], [64, 104, 5], [87, 105, 4], [10, 110, 4], [46, 110, 9], [31, 111, 1], [30, 113, 4], [36, 113, 8], [30, 116, 3], [24, 116, 4], [93, 120, 7], [106, 120, 9], [118, 120, 10], [21, 121, 4], [55, 122, 5], [27, 124, 5], [94, 124, 7], [34, 125, 1], [82, 126, 1], [91, 127, 10], [106, 128, 6], [51, 129, 3], [119, 129, 7], [109, 130, 1], [85, 130, 4], [43, 131, 3], [79, 135, 3], [65, 135, 5], [38, 135, 7], [66, 135, 7], [96, 137, 2], [60, 138, 7], [112, 139, 10], [54, 140, 6], [103, 141, 9], [93, 142, 5], [128, 145, 4], [110, 145, 10], [58, 146, 2], [13